### PR TITLE
Adds coordinate utility unit tests

### PR DESCRIPTION
**Pull request by Crafty15**
- Added unit tests for the coordinate functions 
- uses catch2's BDD style macros. 
- Plz guys it's been 2 years :rofl:

### CODE REVIEW

File: shared/tests/utility/math/docs/cartesianToSpherical.m
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/cartesianToSpherical.m:1]: Ahh, there has been a miscommunication from my end! When I was saying we want to use matlab for our ground truth, what I meant was we should use the matlab builtin functions which do these. So instead of implementing a function which does the same thing on matlab, use their standard library versions [(here's this example)](https://au.mathworks.com/help/matlab/ref/sph2cart.html) for these functions.
Our reasoning is that matlab should certainly have rock solid implementations for these functions (because they're common functions), so we want to match _their_ implementation closely.

My bad for poorly communicating that to you!

File: shared/tests/utility/math/docs/genValsInRange.m
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/genValsInRange.m:5]: If we're not giving matlab a random seed, won't this produce different values each time? I think it would be good if this set the seed, or at least commented, noting the seed which produced the values used in the final tests.

File: shared/tests/utility/math/docs/README.txt
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/README.txt:6]: did `DBL_MIN` give `inf` return values? wtf?

File: shared/tests/utility/math/docs/README.txt
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/README.txt:8]: Noting here as well that it's not that the values aren't possible, it's more that we're assuming they won't happen.

File: shared/tests/utility/math/docs/README.txt
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/README.txt:7]: Name the `functions being tested` so that future code readers know what this refers to.

File: shared/tests/utility/math/docs/README.txt
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/README.txt:14]: Didn't this change to being a percentage threshold, rather than an absolute threshold?

File: shared/tests/utility/math/docs/cartesianToSpherical.m
Comment (KipHamiltons) [reply shared/tests/utility/math/docs/cartesianToSpherical.m:1]: Also, the scripts that remain should be placed in a subfolder of this one I think

File: shared/tests/utility/math/docs/genValsInRange.m
Comment (KipHamiltons) [inline shared/tests/utility/math/docs/genValsInRange.m:3]: I would suggest putting the params on separate lines in this comment. On one line, it's a little harder for me to parse than I think it could be

**Review by KipHamiltons — CHANGES_REQUESTED** [review]
First impressions. I haven't looked at the actual test file yet, but there is some stuff to look at otherwise.

A question as well: what's the thinking behind having the text files duplicating the test data?

File: shared/utility/math/coordinates.hpp
Comment (Crafty15) [reply shared/utility/math/coordinates.hpp:21]: I changed this to resolve a conflict on advice from Ysobel. The .h files were changed to .hpp while i was working on my branch.

Comment (ysims) [reply shared/utility/math/coordinates.hpp:21]: This isn't a change as such, as it is something that didn't get fixed after the merge. You'll want to change it from H to HPP in all cases. Just an annoying thing from the H to HPP change, that is in all branches older than that commit.

File: shared/tests/utility/math/docs/genValsInRange.m
Comment (Crafty15) [reply shared/tests/utility/math/docs/genValsInRange.m:5]: Hmmm, yes. I didn't think of using a seed. I was only accounting for if someone wanted to regen a new set. The extra copies of the test data were included in the docs folder in case of an accidental deletion or something. If we want it to make the same data each time i'm happy to work on it further though.

File: shared/tests/utility/math/docs/cartesianToSpherical.m
Comment (Crafty15) [reply shared/tests/utility/math/docs/cartesianToSpherical.m:1]: That makes sense! Shouldn't be a big change. I'll sort that out.

File: shared/tests/utility/math/docs/genValsInRange.m
Comment (KipHamiltons) [reply shared/tests/utility/math/docs/genValsInRange.m:5]: Since you have it committed to git, there's no need to worry about accidental deletion. And, now that a commit with the data been pushed to the remote repo (and thus to the github servers), the information will live on as long as the repo does, accessible basically forever, through that commit.

File: shared/tests/utility/math/docs/calc_spherToCart.m
Comment (Crafty15) [reply shared/tests/utility/math/docs/calc_spherToCart.m:3]: I think that happens when matlab generates a function template. I'll tidy all this up as i implement the build in functions.

File: shared/tests/utility/math/docs/README.txt
Comment (Crafty15) [reply shared/tests/utility/math/docs/README.txt:14]: Yep. That change happened further down where instead of just comparing the difference between the pre-calc value and the method result in the REQUIRE statement, i used Approx(cart_compare.x()).epsilon(ERROR_THRESHOLD).

Comment (KipHamiltons) [reply shared/tests/utility/math/docs/README.txt:14]: I think this comment in this readme could reflect that we have opted to use a percentage threshold (and also even any reasoning behind that threshold or why we used percentages, if you think that could be useful to future maintainers?)

File: shared/tests/utility/math/docs/README.txt
Comment (Crafty15) [reply shared/tests/utility/math/docs/README.txt:6]: I'm going to have to double check this.

File: shared/tests/utility/math/docs/README.txt
Comment (Crafty15) [reply shared/tests/utility/math/docs/README.txt:1]: Good call! I'll do that.

File: shared/tests/utility/math/docs/cartesianToSpherical.m
Comment (Crafty15) [reply shared/tests/utility/math/docs/cartesianToSpherical.m:1]: For the record, pretty sure you did say "built in functions" but i was thinking of trig functions.

File: .devcontainer.json
Comment (TrentHouliston) [reply .devcontainer.json:4]: Not only that but these changes are a move backwards

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (TrentHouliston) [inline shared/tests/utility/math/TestCoordinates.cpp:43]: For each of these constants, you should try to get them from limits if you can.
e.g. `std::numeric_limits<double>::max()` and `std::numeric_limits<double>::lowest()` etc.
Also when you're doing tests to make sure floating point error hasn't grown enough instead of using a fixed number like 1e-6 for the error threshold, you should be checking against machine epsilon.

Have a look at https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon
Basically, the error you should expect in a set of floating-point operations is related to the size of the two numbers, as well as how many operations you are doing on them.

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (TrentHouliston) [inline shared/tests/utility/math/TestCoordinates.cpp:48]: If you have large sets of data like this in tests, it's generally best practice to separate the "logic" from the "data".
Put these constant sets into other files and include them. That way the logic of the test is clear in one file without getting hidden with large amounts of data like this.

File: shared/tests/utility/math/docs/README.txt
Comment (TrentHouliston) [reply shared/tests/utility/math/docs/README.txt:1]: If you want to see what markdown is supported within here you should look at github flavoured markdown, as that's what it will be rendered with https://github.github.com/gfm/

File: shared/tests/utility/math/docs/README.txt
Comment (TrentHouliston) [reply shared/tests/utility/math/docs/README.txt:6]: The extremely large and small numbers will often cause numerical instability. I'm not sure it makes sense to use them in lots of these tests as you will get nonsensical results. You should mainly make sure that values that are within our domain and range end up with sensible answers.

File: shared/tests/utility/math/docs/README.txt
Comment (TrentHouliston) [reply shared/tests/utility/math/docs/README.txt:14]: Read the information about https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon which will help you write about how you should be testing. It shouldn't be a percentage error or an absolute error per say. It should be an error based on the machine epsilon (which is based on the relative magnitude of the numbers in the operation)

File: shared/tests/utility/math/docs/calc_spherToCart.m
Comment (TrentHouliston) [reply shared/tests/utility/math/docs/calc_spherToCart.m:9]: Our vscode should be putting that in if you save the file in vscode :)

File: shared/tests/utility/math/docs/cart_coords_in_cpp.txt
Comment (TrentHouliston) [inline shared/tests/utility/math/docs/cart_coords_in_cpp.txt:1]: What is the purpose of this file?
I'd recommend either making it a csv file or not having it at all.
likewise for the others

File: shared/tests/utility/math/docs/spher_coords_cpp.txt
Comment (TrentHouliston) [inline shared/tests/utility/math/docs/spher_coords_cpp.txt:1]: I'd recommend removing these files as they clutter up the repo and don't add much. The scripts that made them are an important part. In general, you don't ever want to commit anything that was generated as it's fragile.

**Review by TrentHouliston — COMMENTED** [review]
Some review comments

File: shared/tests/utility/math/docs/README.txt
Comment (KipHamiltons) [reply shared/tests/utility/math/docs/README.txt:6]: This begs the question - how do we define or find out what our domain and range are for a given module?

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (KipHamiltons) [reply shared/tests/utility/math/TestCoordinates.cpp:43]: two questions I have, having read your comment and that page:

1. How would you suggest to take into account how many operations are being done on the floats?
2. In that cppreference example, it has multiplies by a ULP==2. Looking at the [wiki page for ULP](https://en.wikipedia.org/wiki/Unit_in_the_last_place), it has examples which are more like 2^{-51} or 2^{-53}, which makes a lot of sense based on that definition of ULP. Using 2^{-51} where cppreference has used what they're calling `ulp` would certainly cause underflow though, because machine epsilon * 2^{-53} will obviously underflow. Should we disregard that term from the calculation entirely, or simply always use "2" as in their example, or some other approach entirely? It seems like I am probably misunderstanding what cppreference means when it says `ulp`

File: shared/tests/utility/math/docs/cart_coords_in_cpp.txt
Comment (KipHamiltons) [reply shared/tests/utility/math/docs/cart_coords_in_cpp.txt:1]: Would it be worth making a utility for parsing csv files of test data? This will probably be a recurring theme as we develop more test files

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (TrentHouliston) [reply shared/tests/utility/math/TestCoordinates.cpp:43]: Sometimes it's hard to know exactly how many operations are being done on a float. But what you'll see is that each operation will have a particular error in regards to units in the last place. Most of the time the error will be <= 0.5 ULP. If you look up documentation you can find out what exactly the expected ULP error is for any operation you're doing.

As for what it means in general since you're dealing with floating-point numbers there isn't a single "error" value that you can reliably use. You can think of floating-point numbers being scientific notation with a particular number of decimal places. So every time you do a maths operation the amount of error you gain is related to the relative magnitudes of numbers. So as an example, if you had a "float" that held 5 significant figures you might have 5.0000e10 So checking your error is less than 1e-6 doesn't make sense as the smallest possible number difference would be 5.0001e10. 1e-6 is smaller than the smallest increment you could do to that number. That's what the machine epsilon is designed to help represent, the smallest "step" you can possibly take based on the magnitude of the numbers involved. So when you're checking for error and how accurate your calculations are, you should use epsilon in order to work out how many units in the last place you want to be accurate to. From our example, that's equivalent to how many significant figures you want to be accurate too. Because you're dealing with binary though it's more like how many bits of error you want to be able to have.

This is also why if you're ever doing some maths operations and you have some large numbers and small numbers, sometimes it's beneficial to rearrange your functions so you only ever do operations on numbers of similar magnitude because large and small numbers together can result in numerical instability due to the operation being unable to represent the small changes in a large number.

File: shared/tests/utility/math/docs/cart_coords_in_cpp.txt
Comment (TrentHouliston) [reply shared/tests/utility/math/docs/cart_coords_in_cpp.txt:1]: It might be a good idea, the hardest part of all this is where to put it all. I'm not convinced that the doc folder belongs here with the code in the first place

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (TrentHouliston) [reply shared/tests/utility/math/TestCoordinates.cpp:43]: Also as one more example as to where this can get weird. Imagine you were testing a `sin` function and you decided that 1e-6 was a good threshold for checking. Because of how ULP works, if you stuck a number that was smaller than 1e-6 into the sin function (say 1-e10), your error is much much lower than 1e-6 because the relative scales of the two numbers is smaller. Likewise, when you stick a big number into sin your error will be larger

Comment (KipHamiltons) [reply shared/tests/utility/math/TestCoordinates.cpp:43]: Roger that. We'll roll with 2 for now 👍

Comment (KipHamiltons) [reply shared/tests/utility/math/TestCoordinates.cpp:43]: wouldn't that sine example be covered by the `std::fabs(input)` term in their example though?

Comment (TrentHouliston) [reply shared/tests/utility/math/TestCoordinates.cpp:43]: So for a more concrete example if you had a ground truth for v for `sin(1e-10) == v`, your error in those two would be much less than 1e-6. But then if you did `sin(5) == v` you could have a much higher error as the magnitude of the numbers is larger. When you're testing floating point error all you have is ULP because smaller numbers have the same number of ULP error even though they have smaller absolute error.

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (Bidski) [inline shared/tests/utility/math/TestCoordinates.cpp:967]: 99% sure that if you use [fmt ostream](https://fmt.dev/latest/api.html#ostream-api) you cant print `Eigen` vectors and matrices directly (might need to add a transpose as well)
```suggestion
        INFO(fmt::format("Input: {}", cart_input.tranpose()));
```

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (Bidski) [inline shared/tests/utility/math/TestCoordinates.cpp:996]:
```suggestion
        Eigen::Vector3d spher_input   = spher_coords[i];
        Eigen::Vector3d spher_compare = spherToCart_results[i];
        Eigen::Vector3d cart_result   = Eigen::Vector3d::Zero();
```

File: shared/tests/utility/math/docs/cartesianToSpherical.m
Comment (Bidski) [reply shared/tests/utility/math/docs/cartesianToSpherical.m:1]: Is there a Python implementation of these functions? Considering the majority of our code base is C++/Python it would be nicer if we could stick to those languages as much as possible

File: shared/utility/math/coordinates.hpp
Comment (Bidski) [reply shared/utility/math/coordinates.hpp:39]: I would actually duplicate this comment for each offending function. Make it explicit with maybe a pre-/post-condition style comment for each function

**Review by Bidski — CHANGES_REQUESTED** [review]

File: shared/tests/utility/math/docs/cartesianToSpherical.m
Comment (KipHamiltons) [reply shared/tests/utility/math/docs/cartesianToSpherical.m:1]: I have found a python implementation in `astropy` - I think it uses another convention though. That means the results it gives would need to be permuted before we used them, which is fine, but certainly worth noting if that's what you think we should go for.

File: shared/utility/math/coordinates.hpp
Comment (Crafty15) [reply shared/utility/math/coordinates.hpp]: Removed these comments completely. It was early days so i was explaining things to myself. Probably TMI i think.

File: shared/tests/utility/math/docs/README.txt
Comment (Crafty15) [reply shared/tests/utility/math/docs/README.txt]: Removed TMI comments

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (Crafty15) [reply shared/tests/utility/math/TestCoordinates.cpp]: Good call. We implemented these changes promptly.....

File: shared/utility/math/coordinates.hpp
Comment (KipHamiltons) [reply shared/utility/math/coordinates.hpp:76]: We still don't want to change this :P

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (KipHamiltons) [inline shared/tests/utility/math/TestCoordinates.cpp:78]: Given that these can be printed on success, it's probably worth changing these `INFO`'s which reference failure as if it did fail.

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (KipHamiltons) [inline shared/tests/utility/math/TestCoordinates.cpp:116]: Same here

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (KipHamiltons) [inline shared/tests/utility/math/TestCoordinates.cpp:62]: I don't quite understand this comment. What do you mean by "at the epsilon value of the expected results"? I piece together you're referring to the return value of the `Approx(<thing>).epsilon(ERROR_THRESHOLD)`, but calling it the `epsilon value of the expected results` confuses me. Naming these variables `epsilon_<letter>` confuses me too, since catch's name for the `epsilon` function itself seems like a misnomer (on catch's behalf) to me.

**Review by KipHamiltons — COMMENTED** [review]
This is really clean! Nice job. 

Sidenote: We talked a while ago about adding a readme or comment somewhere nearby to test data which detailed how inputs were generated and how the ground truth was generated/checked, so that future people could verify, regenerate, or build upon it. Do you think we could include something like that here?

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (Crafty15) [reply shared/tests/utility/math/TestCoordinates.cpp:62]: Yeah i don't understand it either. Looks like i deleted a line in the middle or something.

File: shared/utility/math/coordinates.hpp
Comment (Crafty15) [reply shared/utility/math/coordinates.hpp:76]: I can't even remember why. Just to clarify, HPP, right?

File: shared/tests/data/CoordinateTests.yaml
Comment (Crafty15) [reply shared/tests/data/CoordinateTests.yaml:213]: > This is really clean! Nice job.
> 
> Sidenote: We talked a while ago about adding a readme or comment somewhere nearby to test data which detailed how inputs were generated and how the ground truth was generated/checked, so that future people could verify, regenerate, or build upon it. Do you think we could include something like that here?

Yeah for sure. I'll whip something up

Comment (Crafty15) [reply shared/tests/data/CoordinateTests.yaml:213]: > 

Yeah something didn't like them. I think it was the yaml cpp. I've completely removed them.

File: shared/utility/math/coordinates.hpp
Comment (KipHamiltons) [reply shared/utility/math/coordinates.hpp:76]: Yep. It's the same as the filename (basically ;P )

**Comment by KipHamiltons** [comment]
Once you add in how you generated the test data (and fix that macroguard ;p ), this LGTM :grin:

**Comment by ysims** [comment]
@Crafty15 did you wanna fix this up and convert to non-draft PR? Looks like it's pretty much done.

**Comment by Crafty15** [comment]
> @Crafty15 did you wanna fix this up and convert to non-draft PR? Looks like it's pretty much done.

Will do

File: shared/tests/data/README.md
Comment (ysims) [inline shared/tests/data/README.md:5]: Should specify which tests/config files these apply to.

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (ysims) [inline shared/tests/utility/math/TestCoordinates.cpp:17]:
```suggestion
 * Copyright 2022 NUbots <nubots@nubots.net>
```

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (ysims) [inline shared/tests/utility/math/TestCoordinates.cpp:52]: Might be good to specify the number of points in the yaml file instead of hardcoding here? or use an `std::vector` if possible?

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (ysims) [inline shared/tests/utility/math/TestCoordinates.cpp:96]: and here

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (KipHamiltons) [reply shared/tests/utility/math/TestCoordinates.cpp:52]: Just for my curiosity - what's the advantage of that?

Comment (ysims) [reply shared/tests/utility/math/TestCoordinates.cpp:52]: new set of data?

Comment (KipHamiltons) [reply shared/tests/utility/math/TestCoordinates.cpp:52]: IIRC, the blame is on me for using an array here. I guess I figured that if you know the size at compile time and it doesn't change, it should be an array, but (at the risk of speaking for Liam here) I would be totally fine with vector here  if you think that's better (I don't think you can use a std::array with a dynamic size/a size only found at runtime).

Comment (Crafty15) [reply shared/tests/utility/math/TestCoordinates.cpp:52]: Doesn't bother me either way. Might be cleaner with a vector

File: shared/tests/utility/math/TestCoordinates.cpp
Comment (Crafty15) [reply shared/tests/utility/math/TestCoordinates.cpp:96]: lol, i forgot to scroll down

**Review by TrentHouliston — APPROVED** [review]

**Review by ysims — APPROVED** [review]
Seems good

**Review by KipHamiltons — APPROVED** [review]

**Review by Bidski — CHANGES_REQUESTED** [review]
You should add an extra test that randomly generates cartesian coordinates then tries to round trip them (`cart -> spherical -> cart -> spherical`) and then check that the two `cart`s are equal and the two `spherical` are equal. This will let us know that the functions are inverses of each other

**Pull request merged**
